### PR TITLE
sgd.py: why add alpha*weights in gradient step?

### PR DESCRIPTION
**Pull request by anuzis**
Reviewing SGD implementations for my own learning, this is one of the more clear examples I've come across, but I'm not sure why it's adding alpha*weights instead of multiplying. Is this a bug?  If not, appreciate any clarification on the approach.

p.s. I don't actually suggest this as a pull request, but am not sure how best to communicate about this code. New-ish to github.  Thanks!  -Michael

### CODE REVIEW

**Comment by dsullivan7** [comment]
Hey, good question. The addition of alpha \* weights is a regularization step. There are more advanced regularization steps such as L1 and L2 which you can read about here: http://research.microsoft.com/pubs/192769/tricks-2012.pdf and here: http://www.deeplearning.net/tutorial/gettingstarted.html. The regularization step implemented here is pretty simple, but if you are just trying to learn, then you might just ignore it because alpha is set to 0.0 on default. If you are interested I also have a cpp implementation which is even simpler: https://github.com/dsullivan7/cpp_sgd/blob/master/sgd.cpp. Let me know if you have any other questions.

**Comment by m4thnerd** [comment]
Yes -- got it working! Glad to finally understand how the gradient is calculated. Thanks for the help.

While troubleshooting, it appears the Static learning_rates_slow class degenerated via:
- self.eta0 \* gradient

This seemed to get it working:
- self.eta0 \* gradient \* (1 / float(X.shape[0]))

Thanks again. -m
